modules/taboolaBidAdapter.js

```diff
@@ -0,0 +1,250 @@
+      }
+    }
+  },
+  getCookieDataByKey(cookieData, key) {
+    const [, value = ''] = cookieData.split(`${key}=`)
+    return value;
+  },
+  getFromLocalStorage() {
+    const {hasLocalStorage, localStorageIsEnabled, getDataFromLocalStorage} = userData.storageManager;
+
+    if (hasLocalStorage() && localStorageIsEnabled()) {
+      return getDataFromLocalStorage(STORAGE_KEY);
+    }
+  },
+  getFromTRC() {
+    return window.TRC ? window.TRC.user_id : 0;
+  }
+}
+
+export const internal = {
+  getPageUrl: (refererInfo = {}) => {
+    if (refererInfo.canonicalUrl) {
+      return refererInfo.canonicalUrl;
+    }
+
+    if (config.getConfig('pageUrl')) {
+      return config.getConfig('pageUrl');
+    }
+
+    try {
+      return getWindowTop().location.href;
+    } catch (e) {
+      return getWindowSelf().location.href;
+    }
+  },
+  getReferrer: (refererInfo = {}) => {
+    if (refererInfo.referer) {
+      return refererInfo.referer;
+    }
+
+    try {
+      return getWindowTop().document.referrer;
+    } catch (e) {
+      return getWindowSelf().document.referrer;
+    }
+  }
+}
+
+export const spec = {
+  supportedMediaTypes: [BANNER],
+  gvlid: GVLID,
+  code: BIDDER_CODE,
+  isBidRequestValid: (bidRequest) => {
+    return !!(bidRequest.sizes &&
+           bidRequest.params &&
+           bidRequest.params.publisherId &&
+           bidRequest.params.tagId);
+  },
+  buildRequests: (validBidRequests, bidderRequest) => {
+    const [bidRequest] = validBidRequests;
+    const {refererInfo, gdprConsent = {}, uspConsent} = bidderRequest;
+    const {bcat = [], badv = [], publisherId} = bidRequest.params;
+    const site = getSiteProperties(bidRequest.params, refererInfo);
+    const device = {ua: navigator.userAgent};
+    const imps = getImps(validBidRequests);
+    const user = {
+      buyeruid: userData.getUserId(gdprConsent, uspConsent),
+      ext: {}
+    };
+    const regs = {
+      coppa: 0,
+      ext: {}
+    };
+
+    if (gdprConsent.gdprApplies) {
+      user.ext.consent = bidderRequest.gdprConsent.consentString;
+      regs.ext.gdpr = 1;
+    }
+
+    if (uspConsent) {
+      regs.ext.us_privacy = uspConsent;
+    }
+
+    if (config.getConfig('coppa')) {
+      regs.coppa = 1
+    }
+
+    const request = {
+      id: bidderRequest.auctionId,
+      imp: imps,
+      site,
+      device,
+      source: {fd: 1},
+      tmax: bidderRequest.timeout,
+      bcat,
+      badv,
+      user,
+      regs
+    };
+
+    const url = [END_POINT_URL, publisherId].join('?pid=');
+
+    return {
+      url,
+      method: 'POST',
+      data: JSON.stringify(request),
+      bids: validBidRequests
+    };
+  },
+  interpretResponse: (serverResponse, {bids}) => {
+    if (!bids) {
+      return [];
+    }
+
+    const {bidResponses, cur: currency} = getBidResponses(serverResponse);
+
+    if (!bidResponses) {
+      return [];
+    }
+
+    return bids.map((bid, id) => getBid(bid.bidId, currency, bidResponses[id])).filter(Boolean);
```

Comment:
this id is actually index, is this how we want it to be?

Reply:
yes.
here is a real request sample. based on increment index:
``` json
"imp": [
    {
      "id": "1",
      "banner": {
        "api": [],
        "battr": [],
        "w": 300,
        "h": 250,
        "pos": 0,
        "topframe": 0,
        "ext": {
          "sizes": [
            {
              "w": 300,
              "h": 250,
              "fmt": [
                15
              ]
            }
          ]
        }
      },
```